pkg/controller/tidbcluster/pod_control.go

```diff
@@ -0,0 +1,318 @@
+	}
+}
+
+// processNextWorkItem dequeues items, processes them, and marks them done. It enforces that the syncHandler is never
+// invoked concurrently with the same key.
+func (c *PodController) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+	result, err := c.sync(key.(string))
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("TidbCluster pod: %v, sync failed %v, requeuing", key.(string), err))
+		c.queue.AddRateLimited(key)
+	} else {
+		if result.RequeueAfter > 0 {
+			c.queue.AddAfter(key, result.RequeueAfter)
+		} else if result.Requeue {
+			c.queue.AddRateLimited(key)
+		} else {
+			c.queue.Forget(key)
+		}
+	}
+	return true
+}
+
+func (c *PodController) sync(key string) (reconcile.Result, error) {
+	ns, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	pod, err := c.deps.PodLister.Pods(ns).Get(name)
+	if errors.IsNotFound(err) {
+		klog.Infof("pod has been deleted %v", key)
+		return reconcile.Result{}, nil
+	}
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	pod = pod.DeepCopy()
+
+	// labels of tikv:
+	// app.kubernetes.io/component=tikv,app.kubernetes.io/instance=db1369682775200135879,app.kubernetes.io/managed-by=tidb-operator ...
+	managedBy := pod.Labels[label.ManagedByLabelKey]
+	if managedBy != "tidb-operator" {
+		return reconcile.Result{}, nil
+	}
+
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished syncing TidbCluster pod %q (%v)", key, time.Since(startTime))
+	}()
+
+	component := pod.Labels[label.ComponentLabelKey]
+	ctx := context.Background()
+	switch component {
+	case "tikv":
+		tcName := pod.Labels[label.InstanceLabelKey]
+		tc, err := c.deps.TiDBClusterLister.TidbClusters(ns).Get(tcName)
+		if err != nil {
+			return reconcile.Result{}, perrors.Annotatef(err, "failed to get tc %q", tcName)
+		}
+		tc = tc.DeepCopy()
+
+		return c.syncTiKVPod(ctx, pod, tc)
+	default:
+		return reconcile.Result{}, nil
+	}
+}
+
+func (c *PodController) getPDClient(tc *v1alpha1.TidbCluster) pdapi.PDClient {
+	if c.testPDClient != nil {
+		return c.testPDClient
+	}
+
+	pdClient := controller.GetPDClient(c.deps.PDControl, tc)
+	return pdClient
+}
+
+func (c *PodController) syncTiKVPod(ctx context.Context, pod *corev1.Pod, tc *v1alpha1.TidbCluster) (reconcile.Result, error) {
+	value, ok := pod.Annotations[v1alpha1.EvictLeaderAnnKey]
+
+	if ok {
+		switch value {
+		case v1alpha1.EvictLeaderValueNone:
+		case v1alpha1.EvictLeaderValueDeletePod:
+		default:
+			klog.Warningf("ignore unknown value %q of annotation %q", value, v1alpha1.EvictLeaderAnnKey)
+			return reconcile.Result{}, nil
+		}
+	}
+
+	if ok {
+		evictStatus := &v1alpha1.EvictLeaderStatus{
+			PodCreateTime: pod.CreationTimestamp,
+			Value:         value,
+		}
+		nowStatus := tc.Status.TiKV.EvictLeader[pod.Name]
+		if nowStatus == nil || *nowStatus != *evictStatus {
+			if tc.Status.TiKV.EvictLeader == nil {
```

Comment:
`if tc.Status.TiKV.EvictLeader == nil {` then L224 will panic?

Reply:
no.
all read operation against a nil map has the same behavior as an empty map!!!